### PR TITLE
lightgbm v4.1.0

### PR DESCRIPTION
**Pull request by lorepirri**
[PKG-3281] lightgbm v4.1.0

- upstream: https://github.com/Microsoft/LightGBM/tree/v4.1.0
- dependency files: 
    - https://github.com/microsoft/LightGBM/blob/v4.1.0/python-package/pyproject.toml
- conda-forge: https://github.com/conda-forge/lightgbm-feedstock
- pypi: https://pypi.org/project/lightgbm
    
**Destination channel:** `defaults`

**Changes**
- refactor recipe, fix linter
- update to `4.1.0`

**Depends on:**
- `scikit-build-core>=0.4.4` AnacondaRecipes/scikit-build-core-feedstock#1

[PKG-3281]: https://anaconda.atlassian.net/browse/PKG-3281?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -25,25 +24,30 @@ requirements:
...
     - scipy
-    - scikit-learn !=0.22.0
-    - _openmp_mutex      # [linux]
 
 test:
   imports:
     - lightgbm
   requires:
     - pip
+    - scikit-learn !=0.22.0
```
Comment (cbouss) [inline recipe/meta.yaml]:
```suggestion
    - scikit-learn
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -25,25 +24,30 @@ requirements:
...
-    - llvm-openmp 14.0.6  # [osx]
-    # libgomp pulled in through gcc setup
-    # - libgomp 11.2.0      # [linux]
+    - llvm-openmp  # [osx]
+    - libgomp      # [linux]
   run:
+    - blas=*=openblas # [osx and x86_64]  
     - python
     - numpy
     - scipy
```
Comment (cbouss) [inline recipe/meta.yaml:42]:
```suggestion
    - scipy
  run_constrained: 
    - scikit-learn !=0.22.0
    - pandas>=0.24.0
    - dask >=2.0.0
```
https://github.com/microsoft/LightGBM/blob/v4.1.0/python-package/pyproject.toml#L35

File: recipe/meta.yaml [hunk 3]
```diff
@@ -25,25 +24,30 @@ requirements:
...
     - setuptools
     - wheel
-    - llvm-openmp 14.0.6  # [osx]
-    # libgomp pulled in through gcc setup
-    # - libgomp 11.2.0      # [linux]
+    - llvm-openmp  # [osx]
+    - libgomp      # [linux]
   run:
+    - blas=*=openblas # [osx and x86_64]  
     - python
```
Comment (cbouss) [inline recipe/meta.yaml:39]:
```suggestion
    - python
    - dataclasses # [py<37]
```
for completeness

File: recipe/meta.yaml (hunk 1 above)
Comment (lorepirri) [reply recipe/meta.yaml]: ah got it, it will be constrained in the `run_constrained`. Thank you

**Review by cbouss — APPROVED** [review]

File: recipe/meta.yaml [hunk 4]
```diff
@@ -25,25 +24,35 @@ requirements:
...
+    # This basically forces to pick scipy (and any other suitable packages in the chain, like numpy) 
+    # that was built with openblas rather than mkl. Openblas typically pairs with llvm-openmp and mkl with intel-openmp on osx-64. 
+    # By default, mkl variants are picked. By picking the variant that uses openblas rather 
+    # than mkl, intel-openmp will not be pulled in and so you will not have llvm-openmp / intel-openmp conflict. 
+    - blas=*=openblas # [osx and x86_64]
+    - scikit-build-core >=0.4.4
     - python
     - pip
     - setuptools
     - wheel
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml]:
```suggestion
```

**Review by JeanChristopheMorinPerso — APPROVED** [review]

**Pull request merged**